accounts/abi/bind/backends/simulated.go

```diff
@@ -398,39 +403,62 @@ func (b *SimulatedBackend) EstimateGas(ctx context.Context, call ethereum.CallMs
 	cap = hi
 
 	// Create a helper to check if a gas allowance results in an executable transaction
-	executable := func(gas uint64) bool {
+	executable := func(gas uint64) (bool, *core.ExecutionResult, error) {
 		call.Gas = gas
 
 		snapshot := b.pendingState.Snapshot()
-		_, _, failed, err := b.callContract(ctx, call, b.pendingBlock, b.pendingState)
+		res, err := b.callContract(ctx, call, b.pendingBlock, b.pendingState)
 		b.pendingState.RevertToSnapshot(snapshot)
 
-		if err != nil || failed {
-			return false
+		if err != nil {
+			if err == core.ErrIntrinsicGas {
+				return true, nil, nil // Special case, raise gas limit
+			}
+			return true, nil, err // Bail out
 		}
-		return true
+		return res.Failed(), res, nil
 	}
 	// Execute the binary search and hone in on an executable gas limit
 	for lo+1 < hi {
 		mid := (hi + lo) / 2
-		if !executable(mid) {
+		failed, _, err := executable(mid)
+
+		// If the error is not nil(consensus error), it means the provided message
+		// call or transaction will never be accepted no matter how much gas it is
+		// assigned. Return the error directly, don't struggle any more
+		if err != nil {
+			return 0, err
+		}
+		if failed {
 			lo = mid
 		} else {
 			hi = mid
 		}
 	}
 	// Reject the transaction as invalid if it still fails at the highest allowance
 	if hi == cap {
-		if !executable(hi) {
-			return 0, errGasEstimationFailed
+		failed, result, err := executable(hi)
+		if err != nil {
+			return 0, err
+		}
+		if failed {
+			if result != nil && result.Err != vm.ErrOutOfGas {
+				errMsg := fmt.Sprintf("always failing transaction (%v)", result.Err)
```

Comment:
@rjl493456442 This currently produces the following out put for me:
failed to estimate gas needed: always failing transaction (execution reverted) (0x08c379a...000)
Which means users would have to parse the error string by hand whenever they want to check the reason, maybe we could add an error type with an attribute "reason" or something.

Reply:
Yes, so I am asking what's the format we should use in API level.

Now the format is `always failing transaction  (ERROR_TYPE) (REASON_STRING)`

Btw we can't provide the decoded reason since we don't have the abi to decode it.

Reply:
I would propose an error type like this 
```
type failedTxError struct {
error_type string
reason string
}
```
So user would not have to manually parse the "always failing tx..." string.

Additionally I think we could decode the reason, as reasons are encoded as if they were a call to `Error(String)` see https://solidity.readthedocs.io/en/latest/control-structures.html#revert

Decoding would become: 
```
if bytes.Equal(rval[:4], crypto.Keccak256([]byte("Error(string)"))[:4]) {
		str, _ := abi.NewType("string", "", nil)
		errorDesc, err := abi.Arguments{{Type: str}}.UnpackValues(rval[4:])
		if err != nil {
			return nil, err
		}
		panic(errorDesc[0].(string))
		return nil, errors.New(errorDesc[0].(string))
	}
```

Reply:
@MariusVanDerWijden Yes you are right. I always think it's user-specific, but the argument for `revert` or `require` is string only. Thanks for pointing out here!

Reply:
@MariusVanDerWijden Yes you are right. I always think it's user-specific, but the argument for `revert` or `require` is string only. Thanks for pointing out here!